clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -373,18 +374,101 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
 
   exchange(uint8_t *local_memory) : _local_memory(local_memory) {}
 
+  static int adjust_by_padding(int offset) {
+
+    if constexpr (INSERT_PADDING) {
+      offset = detail::shr_add(offset, LOG_LOCAL_MEMORY_BANKS, offset);
+    }
+    return offset;
+  }
+  template <typename Item>
+  struct stripedToBlockedFunctor {
+
+    int forward_offset(Item item, int i) {
+      int fw_offset = int(i * item.get_local_range(2) *
+                          item.get_local_range(1) * item.get_local_range(0)) +
+                      item.get_local_id(0);
+      return adjust_by_padding(fw_offset);
+    }
+
+    int reverse_offset(Item item, int i) {
+      int rv_offset = int(item.get_local_id(0) * VALUES_PER_THREAD) + i;
+      return adjust_by_padding(rv_offset);
+    }
+  };
```

Comment:
This is headed in the right direction I think, but still needs a bit of a tweak.

Here you name one function as `forward_offset` and the other as `reverse_offset`.  Those names only make sense in the context of the striped to blocked operation, but they are also used for the blocked to striped operation.  I don't see any reason to tie them together, it just makes things harder.

If it were me, I would just have two different functors, which each overload the `operator()`, `get_striped_offset` and `get_blocked_offset`.  Then you can send these directly in to `helper_exchange`, without needing a separate struct and member function.  You can just call them with the `operator()` within `helper_exchange`.

Then, if you want, I think this would just require one more functor to implement the `get_scatter_offset` which uses an map to get the index.

Reply:
Made the necessary changes , could you please review @danhoeflinger

Reply:
Since some methods ( scatter -> block/scatter->striped) rely on ranks, the helper_exchange has an overloaded counterpart with ranks arg .  The scatter -> striped part is also added for review.

Reply:
Continuing to look better.  I think you can reuse the same `helper_exchange` for both cases if you just initialize the scatter functor with the `ranks` array in its constructor.

Reply:
Thanks for the suggestion @danhoeflinger  .Made the changes inside the constructor .